Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GameController.cs

```diff
@@ -0,0 +1,337 @@
+        [InputControl(name = "rightShoulder", template = "Button", bit = (uint)KeyCode.BUTTON_R1)]
+        [InputControl(name = "start", template = "Button", bit = (uint)KeyCode.BUTTON_START)]
+        [InputControl(name = "select", template = "Button", bit = (uint)KeyCode.BUTTON_SELECT)]
+        [InputControl(name = "dpadButtons", template = "Dpad")]
+        [InputControl(name = "dpadButtons/up", template = "Button", bit = (uint)KeyCode.DPAD_UP)]
+        [InputControl(name = "dpadButtons/right", template = "Button", bit = (uint)KeyCode.DPAD_RIGHT)]
+        [InputControl(name = "dpadButtons/down", template = "Button", bit = (uint)KeyCode.DPAD_DOWN)]
+        [InputControl(name = "dpadButtons/left", template = "Button", bit = (uint)KeyCode.DPAD_LEFT)]
+        public fixed uint buttons[(kMaxAndroidButtons + 31) / 32];
+
+        private const uint kAxisOffset = sizeof(uint) * (uint)((kMaxAndroidButtons + 31) / 32);
+        [InputControl(name = "leftTrigger", template = "Button", format = "FLT", offset = (uint)Axis.LTRIGGER * sizeof(float) + kAxisOffset)]
+        [InputControl(name = "rightTrigger", template = "Button", format = "FLT", offset = (uint)Axis.RTRIGGER * sizeof(float) + kAxisOffset)]
+        [InputControl(name = "leftStick", template = "Stick", format = "VC2F")]
+        [InputControl(name = "leftStick/x", format = "FLT", offset = (uint)Axis.X * sizeof(float))]
+        [InputControl(name = "leftStick/y", format = "FLT", offset = (uint)Axis.Y * sizeof(float) )]
+        [InputControl(name = "rightStick", template = "Stick", format = "VC2F")]
+        [InputControl(name = "rightStick/x", format = "FLT", offset = (uint)Axis.Z * sizeof(float))] 
+        [InputControl(name = "rightStick/y", format = "FLT", offset = (uint)Axis.RZ * sizeof(float))]
+        [InputControl(name = "dpadAxes", template = "Dpad")]
+        [InputControl(name = "dpadAxes/up", template = "Button", format = "FLT", offset = (uint)Axis.HAT_Y * sizeof(float))]
+        [InputControl(name = "dpadAxes/right", template = "Button", format = "FLT", offset = (uint)Axis.HAT_X * sizeof(float))]
+        [InputControl(name = "dpadAxes/down", template = "Button", format = "FLT", offset = (uint)Axis.HAT_Y * sizeof(float))]
+        [InputControl(name = "dpadAxes/left", template = "Button", format = "FLT", offset = (uint)Axis.HAT_X * sizeof(float))]
+        public fixed float axis[kMaxAndroidAxes];
+
+        public FourCC GetFormat()
+        {
+            return kFormat;
+        }
+    }
+
+    [InputTemplate(stateType = typeof(GameControllerState))]
+    public class GameController : Gamepad
+    {
+        protected override void FinishSetup(InputControlSetup setup)
+        {
+            // TODO: pick dpadAxes or dpadButtons? We don't upfront if dpad is bound to buttons or hat switch
+            //       only when we first receive event for ex., HAT_X or DPAD_UP, we can then assume that one or another should be used

```

Comment:
@Rene-Damm There is this little problem, when game controller is connected I don't know if DPad will be coming via buttons or axes

Reply:
That's an interesting case. It seems like this would need a custom Dpad control which is wired to both the button and the axis state and dynamically decides which one to use.

I do wonder, is this the case on Android even when using an actual gamepad with a normal button layout (i.e. on one gamepad you get buttons and on another gamepad with the same layout you get axes; that'd be super weird) or is it just something that varies between joysticks and gamepads? Can we detect it by device type?

Reply:
Ok, judging from your other comment, that's indeed a pretty messed up setup. Seems like it will indeed require an Android-specific DpadControl which taps both states and dynamically returns one or the other.

One problem we'll likely run into is that the two states don't form a contiguous chunk of memory. It's a bit over here and a float over there. Likely to cause some issues in combination with input actions. But let's see.

Reply:
Actually, I am stupid... there is a way, when device is connected, you can get a list of available motion event types from device, so in this case  if MOTION_HAT_X , MOTION_HAT_Y are available, we know that DPAD will be coming from there, so I'll guess I'll put that info into capabilities of device